From 8dac2b2b6c7196fa299ce99f2303683d9db9131b Mon Sep 17 00:00:00 2001
From: Henry Wang <Henry.Wang@arm.com>
Date: Sun, 9 Oct 2022 02:24:42 +0000
Subject: [PATCH] xen/arm,sysctl: Add SLC-related PSR sysctl subops

Users need to get the system level information for SLC through the
xl-psr interfaces. Introduce sysctl subops
`XEN_SYSCTL_PSR_CMT_get_slc_size` and `XEN_SYSCTL_PSR_get_slc_info`
for this purpose. The two newly introduced sysctl subops will
call related MPAM helpers with type MPAM_CLASS_SLC to fetch the
information accordingly.

Signed-off-by: Henry Wang <Henry.Wang@arm.com>
Upstream-Status: Pending
Issue-Id: SCM-5244
---
 xen/arch/arm/sysctl.c       | 24 ++++++++++++++++++++++++
 xen/include/public/sysctl.h |  2 ++
 2 files changed, 26 insertions(+)

diff --git a/xen/arch/arm/sysctl.c b/xen/arch/arm/sysctl.c
index 55d1ca080974..763f512be44d 100644
--- a/xen/arch/arm/sysctl.c
+++ b/xen/arch/arm/sysctl.c
@@ -59,6 +59,16 @@ long arch_do_sysctl(struct xen_sysctl *sysctl,
 
             break;
         }
+        case XEN_SYSCTL_PSR_CMT_get_slc_size:
+        {
+            unsigned int slc_size = 0;
+
+            mpam_get_cache_size(MPAM_CLASS_SLC, 0, &slc_size);
+            /* in KB unit */
+            sysctl->u.psr_cmt_op.u.data = (slc_size ? slc_size / 1024 : 0);
+
+            break;
+        }
         default:
             sysctl->u.psr_cmt_op.u.data = 0;
             ret = -EOPNOTSUPP;
@@ -90,6 +100,20 @@ long arch_do_sysctl(struct xen_sysctl *sysctl,
             if ( __copy_field_to_guest(u_sysctl, sysctl, u.psr_alloc) )
                 ret = -EFAULT;
             break;
+
+        case XEN_SYSCTL_PSR_get_slc_info:
+            ret = mpam_get_hwinfo(MPAM_CLASS_SLC, 0, &hw_val);
+            if( ret )
+                break;
+
+            sysctl->u.psr_alloc.u.cat_info.cos_max =
+                mpam_get_system_max_partid();
+            sysctl->u.psr_alloc.u.cat_info.cbm_len = hw_val;
+
+            if ( __copy_field_to_guest(u_sysctl, sysctl, u.psr_alloc) )
+                ret = -EFAULT;
+            break;
+
         default:
             ret = -EOPNOTSUPP;
             break;
diff --git a/xen/include/public/sysctl.h b/xen/include/public/sysctl.h
index f1eba784059a..0e3a14a507e5 100644
--- a/xen/include/public/sysctl.h
+++ b/xen/include/public/sysctl.h
@@ -834,6 +834,7 @@ struct xen_sysctl_coverage_op {
 #define XEN_SYSCTL_PSR_CMT_get_l3_cache_size         2
 #define XEN_SYSCTL_PSR_CMT_enabled                   3
 #define XEN_SYSCTL_PSR_CMT_get_l3_event_mask         4
+#define XEN_SYSCTL_PSR_CMT_get_slc_size              5
 struct xen_sysctl_psr_cmt_op {
     uint32_t cmd;       /* IN: XEN_SYSCTL_PSR_CMT_* */
     uint32_t flags;     /* padding variable, may be extended for future use */
@@ -871,6 +872,7 @@ struct xen_sysctl_pcitopoinfo {
 #define XEN_SYSCTL_PSR_get_l3_info               0
 #define XEN_SYSCTL_PSR_get_l2_info               1
 #define XEN_SYSCTL_PSR_get_mba_info              2
+#define XEN_SYSCTL_PSR_get_slc_info              3
 struct xen_sysctl_psr_alloc {
     uint32_t cmd;       /* IN: XEN_SYSCTL_PSR_* */
     uint32_t target;    /* IN */
-- 
2.34.1

